### PR TITLE
Add combinator for adding effects at an index

### PR DESCRIPTION
**Pull request by tek**
opinions?

### CODE REVIEW

**Comment by googleson78** [comment]
I wonder if you can't somehow reuse `Here/There` for this (?) :thinking:

File: src/Polysemy/Internal.hs [hunk 1]
```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
...
+class KnownList l where
+  singList :: SList l
+
+instance KnownList '[] where
+  singList = SEnd
+  {-# INLINE singList #-}
+
+instance KnownList xs => KnownList (x ': xs) where
+  singList = SCons singList
+  {-# INLINE singList #-}
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:614]: We should put all of the `SList` and `KnownList` boilerplate in a different module; I suggest `Polysemy.Internal.Membership`.
Also, replace the existing `KnownList` representation in `Polysemy.Internal.Bundle` with usages of this new one, instead.

File: src/Polysemy/Internal.hs [hunk 2]
```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
...
+instance {-# INCOHERENT #-} (Build l1 s r ix head return, l ~ (e ': l1))
+      => Build l (e ': s) r ix head return where
+  singBuild = SCons (singBuild @_ @l1 @s @r @ix @head @return)
+
+instance {-# INCOHERENT #-} TypeError (InsertAtFailure r ix head return)
+       => Build l s r ix head return where
+  singBuild = undefined
+
+type family InsertAtUnprovidedIndex where
+  InsertAtUnprovidedIndex = TypeError ('Text "insertAt: index must be provided")
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:640]: `Tide` and `Build` were placeholder names, and should be replaced. My suggestions are`KnownListOfLength` and `KnownPrefix`.

`Tide` isn't `insertAt`-specific, and should reside in `Polysemy.Internal.Membership`. A silly thing here is that the only thing that makes `Build` `insertAt` specific is the need to propagate information for error messages. I suppose that can't be helped.

File: src/Polysemy/Internal.hs [hunk 3]
```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
...
+insertAt
+  :: forall n new top r s a
+   . ( Tide n top
+     , WhenStuck n InsertAtUnprovidedIndex
+     , s ~ Append new r, Build new s r n top (Append top s)
+     , Subsume (Append top r) (Append top s))
+  => Sem (Append top r) a
+  -> Sem (Append top s) a
+insertAt =
+  subsume_
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:663]: Remember that `subsume_` won't work here. You need to steal [`weakenMid`](https://hackage.haskell.org/package/in-other-words-0.2.0.0/docs/src/Control.Effect.Internal.KnownList.html#weakenMid) from IOW and use that, instead.

Implementation should be something like this (haven't doublechecked):
```
insertAt = hoistSem $ \u -> hoist (insertAt @n @new @top @r @s) $
  weakenMid @r (singTide @n @top) (singBuild @_ @new @s @r @n @top @(Append top s)) u
```
Some type applications may be redundant.

File: src/Polysemy/Internal.hs [hunk 4]
```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
...
+--
+-- @since TODO
+insertBefore ::
+  ∀ e es r r' a .
+  r' ~ InsertedBefore e es r =>
+  Subsume r r' =>
+  Sem r a ->
+  Sem r' a
+insertBefore = subsume_
+{-# inline insertBefore #-}
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:597]: Given the problems I pointed out with `insertBefore` and `insertAfter`, my opinion remains that we shouldn't include them.

File: polysemy-plugin/test/InsertSpec.hs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import Polysemy.Reader (Reader)
+import Polysemy.State (State)
+import Polysemy.Internal (Sem)
+import Polysemy.Internal (insertAt)
+import Polysemy
+import Test.Hspec
+
+insertSome :: ∀ e1 e2 i r a . Sem (e1 ': e2 ': r) a -> Sem (e1 ': e2 ': Reader i ': r) a
+insertSome = insertAt @2 @'[Reader i]
```
Comment (KingoftheHomeless) [inline polysemy-plugin/test/InsertSpec.hs:11]: Drop `@'[Reader i]`, it shouldn't be necessary.

File: polysemy-plugin/test/InsertSpec.hs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+--   Sem (e1 ': e2 ': r) a ->
+--   Sem (e1 ': e2 ': State Int ': e4 ': Reader i ': r) a
+-- insertAtTest2 = insertAt @2 @'[State Int, e4, Reader i]
+
+
+spec :: Spec
+spec = parallel $ do
+  describe "insert" $ do
+    it "insert" $ do
+      1 `shouldBe` 1
```
Comment (KingoftheHomeless) [inline polysemy-plugin/test/InsertSpec.hs:31]: If we can bear the _headache_, we should have tests for compile-time error messages like the `polysemy-plugin` tests do.

File: src/Polysemy/Internal.hs (hunk 4 above)
Comment (KingoftheHomeless) [reply src/Polysemy/Internal.hs:597]: That is, `insertBefore` and `insertAfter` plays hell with even partially polymorphic effects. This doesn't work:
```
insertAfterTest :: forall i r a. Sem (Reader Int ': Reader i ': r) a -> Sem (Reader Int ': Reader i ': State Int ':  r) a
insertAfterTest = insertAfter @(Reader i) @'[State Int]
```
This is unavoidable as long as we try to use effect equality in order to determine where the new effects should be inserted, making `insertAfter` and `insertBefore` fundamentally flawed.

File: src/Polysemy/Internal.hs [hunk 7]
```diff
@@ -534,6 +540,143 @@ subsumeUsing pr =
...
+--   sem1 = insertAt @2 @[e3, e4] (sem0 :: Sem [e1, e2, e5] a)
+-- @
+--
+-- @since TODO
+insertAt
+  :: forall n new top r s a
+   . ( Tide n top
+     , WhenStuck n InsertAtUnprovidedIndex
+     , s ~ Append new r, Build new s r n top (Append top s)
+     , Subsume (Append top r) (Append top s))
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:659]: No `Subsume`. Begone.

**Review by KingoftheHomeless — COMMENTED** [review]
I realize this is all just placeholder for the snippets of code I threw out, but I figured I could place some signposts as to how we should proceed from here.

File: src/Polysemy/Internal.hs [hunk 8]
```diff
@@ -342,7 +352,7 @@ hoistSem nat (Sem m) = Sem $ \k -> m $ \u -> k $ nat u
 
 
 ------------------------------------------------------------------------------
--- | Introduce an arbitrary number of effects on top of the effect stack. This
+-- | Introduce an arbitrary number of effects on head of the effect stack. This
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:355]: the head

File: src/Polysemy/Internal.hs [hunk 9]
```diff
@@ -464,7 +474,7 @@ raise3Under = hoistSem $ hoist raise3Under . weaken3Under
 
 
 ------------------------------------------------------------------------------
--- | Allows reordering and adding known effects on top of the effect stack, as
+-- | Allows reordering and adding known effects on head of the effect stack, as
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:477]: the head

File: src/Polysemy/Internal.hs [hunk 10]
```diff
@@ -540,6 +550,32 @@ subsumeUsing pr =
...
+--
+-- @since TODO
+insertAt
+  :: forall index inserted head oldTail tail old full a
+   . ( ListOfLength index head
+     , WhenStuck index InsertAtUnprovidedIndex
+     , old ~ Append head oldTail
+     , tail ~ Append inserted oldTail
+     , full ~ Append head tail
+     , InsertAtIndex index head tail oldTail full inserted)
```
Comment (KingoftheHomeless) [inline src/Polysemy/Internal.hs:572]: Good lord... I know there's not much we can do to shorten this type signature, but still.

**Review by KingoftheHomeless — APPROVED** [review]
Finish the rewrite on `Bundle` and this looks good.

File: src/Polysemy/Internal.hs (hunk 10 above)
Comment (tek) [reply src/Polysemy/Internal.hs:572]: well those are basically local type binders…and an error message. Effectively it's just two constraints :upside_down_face:

File: src/Polysemy/Internal.hs (hunk 8 above)
Comment (tek) [reply src/Polysemy/Internal.hs:355]: whoops, that was a _replace all_ :sweat_smile:

Comment (tek) [reply src/Polysemy/Internal.hs:355]: changing it back to the original wording

Comment (tek) [reply src/Polysemy/Internal.hs:355]: maybe _stack_ is a misleading term, given that we're inserting effects in the middle of the list

**Pull request merged**